api/app/src/main/kotlin/packit/service/UserRoleFilterService.kt

```diff
@@ -0,0 +1,153 @@
+package packit.service
+
+import org.springframework.stereotype.Service
+import packit.model.Packet
+import packit.model.Role
+import packit.model.User
+import packit.model.dto.RolesAndUsers
+import packit.security.PermissionChecker
+
+interface UserRoleFilterService
+{
+    fun getRolesAndUsersWithSpecificReadPacketGroupPermission(
+        roles: List<Role>,
+        users: List<User>,
+        packetGroupName: String
+    ): RolesAndUsers
+
+    fun getRolesUsersWithSpecificReadPacketPermission(
+        roles: List<Role>,
+        users: List<User>,
+        packet: Packet
+    ): RolesAndUsers
+
+    fun getRolesAndUsersCantPacketReadGroup(
+        roles: List<Role>,
+        users: List<User>,
+        packetGroupName: String
+    ): RolesAndUsers
+
+    fun getRolesAndUsersCantReadPacket(
+        roles: List<Role>,
+        users: List<User>,
+        packet: Packet
+    ): RolesAndUsers
+
+    fun getRolesAndSpecificUsersCanReadPacket(
```

Comment:
I've lost track of what "Specific" means in this context. In the earlier functions "specific" means eg. `packet.read:foo:<id>` for the `<id>` packet, but here it is applied to the User? Does it mean non-inherited via a non-username role?

Reply:
Yeah second point... Specific means its permissions are particular to that user... i.e not inherited via role